### ISSUE TITLE
Why godot has no api?

### ISSUE THREAD

**Issue by CharlesWoodhill** [godotengine/godot#8065]
**Operating system or device - Godot version:**
all/all

**Issue description:**
Why godot has no api?

Since godot has no api, every project that wants to use godot functionality 
has to write/generate its own godot api, or as in 2nd project (re)write godot fuctionality.

Wouldnt it be simpler to provide an api thats able, to pass dll border?
I am not talkig about one that only supports godot-reflection-systems funcs,
because for example the visual server is not present there.
Also im wondering why godot is not "modularised" since for example visual servers api would support it.

godotCore.dll, godotScenes.dll, visualServer.dll, somesound.dll
godotEditor.exe, godotGame.exe

this would provide much more functionality:
1. One still could statically link into 1 godot.exe file if wanted ( but why :) )
2. It would make the mentioned projects much more easier.
3. (Automated) testing would be much easier, (a small c++ or c# test-program. that doesnt "pollute"
    the godot.exe
4. no more need to always compile whole godot, just compile your module (dll, so, a) 
    with your preferred build system. 
5. no need to integrate a programming language into godot, 
    because its easy to integrate godot into the language (if it supports c integration) like c#,...

and much more :)

**Link to minimal example project:**
https://github.com/GodotNativeTools/dlscript_module
https://github.com/neikeq/GodotSharp

**Comment by Marqin** [comment]
Godot uses C++, and because of name-mangling it's not easy to make C++-API (and doing it as C-API will stripe us from all features that C++ provides). It could be done with `libclang` reflections, but probably there are more important things for devs right now (well, you can make PR).

**Comment by CharlesWoodhill** [comment]
that was my question, why...   is it not wanted, or is there no dev time for it? ( may do it, but not as sole project, since its a core feature)

**Comment by karroffel** [comment]
@CharlesWoodhill Hi. I work on the DLScript module and you've got the wrong impression of how Godot works and *is supposed to work*.

(also [VisualServer](http://docs.godotengine.org/en/stable/classes/class_visualserver.html) is exposed to the reflection system)

There's a difference between "engine-authorative" and "application-authorative" game frameworks. (I don't think those are real terms but I'll explain what I mean).

 - In an "application-authorative" framework you write a program that controlls how rendering and input and stuff like that happens. It uses functionality given by the game framework but it's managing its own main loop.
 - In an "engine-authorative" framework you write scripts or plugins for an already existing system. The engine is a program itself. The engine has the main loop and calls your scripts whenever it thinks it's meaningful.

"application-authorative" is usually just called a "framework". There are many examples for those, SFML, Löve2D, MonoGame, LibGDX.

"engine-authorative" is usually just called a "game engine". Unreal, Godot, Unity, GM.

Godot is a *game engine*, it handles all the hard stuff for you and if you want to use Godot then you need to write code that it can work with. Basically all you interaction with Godot is via the scripting system. Godot uses your scripts to handle game logic etc.

So instead of writing a program that uses Godot functionality you write code that Godot uses. Just having a `visualServer.dll` on its own is useless. This component only makes sense in composition with all the other core stuff that's in Godot.

I think your expectations are wrong because you don't write programs that use Godot. You write scripts that Godot can use. And the DLScript module enables you to use shared libraries as scripts.
I decoupled the cpp implementation of the core types and the class API from DLScript, there *are* in fact two libraries you can use to develop DLScript libraries. 
![image](https://cloud.githubusercontent.com/assets/5209613/24072320/dea6f7ee-0be5-11e7-9e98-f6ecc9f4abac.png)

To sum up: decoupling all those core components of Godot is useless because *they only make sense in composition*.
Having an API to write programs that make use of Godot functionality is useless because Godot components only **work** *in composition of the whole Godot system*, so a standalone application using Godot makes no sense either.





(*and also yes, C++ is a b\*tch*)

**Comment by CharlesWoodhill** [comment]
"application-authoritative"  & "engine-authoritative"  are very good terms :)
but even if i would have that api i wouldnt do my own main loop. i still would use almost everything of godot including the scene system, but i will never use gdscript, because its very limited and not portable if i wanna use an other engine. i have tons of c# code i want to reuse, and that would be much easier with an api. i can use godots main.h funtions to host godot in c# already did it for fun, that way i managed to run godot in an win32-window-panel, surronded by normal .net gui. all that was missing was the api... (if its useful is on another paper :) ) thats why i started to write a wrapper using clang, i stopped it because of neikeqs c# module, i just started to work on it again. because the project isnt continued or low prio, also it only uses mono making debuggin cumbersome, and nothing reaches the vstudio debugger with debug through (c++ <-> c#). but for this u have to use microsoft .net for developing.

i even wanna use gdscript, but only for modding, and of course for one of my ideas "robowars" where the player can script his robots ai :) for this its perfect. 
i think that, only scripting isnt enuf (maybe for some small technically-non-complex games), and c++ is too slow(developing time ofc).
especially because godot is sill young, and doesnt have broad functionalty yet.
(missing terrain is a good example, but thers more)
its just for the case when one writes an module like those 2 projects (and i wanna so similar one).
and one doesnt exclude the other ("application-authoritative"  & "engine-authoritative")

u ofc doin a great job with gdscript-centered layout. 
but this only satisfy artists needs, developing in gdscript is too cumbersome because u loose project overview if getting big, the debugger is... well... not ready :), and code completition (which i normally use to explore the used engine) is not working, and maybe will never work because of variants.

c++ on the other hand is too slow (devtime) and error prone especially in memory management.

still i think an api doesnt hurt. 2 projects already make or made their own one, which is waste of time, so there is need :)

**Comment by CharlesWoodhill** [comment]
"application-authoritative"  & "engine-authoritative"  are very good terms :)
maybe i stressed that decoupling a bit too much... that wasnt really the point.

the point was the api.

also an api doesnt break engine-autorative, 
2 projects already make/made their own, they stick to the engine-authorative way
but still need to write their own godot api...this is waste of time :)

i dont think one need to seperate those to ways when making a game engine,
("application-authoritative"  & "engine-authoritative") both needs can be fulfilled,
no need to decide for only one way.

this for instance, brought me to godot (from unity) because u have the whole engine, and not only those things some devs (or artists) think u need. but i think i was wrong :)

**Comment by karroffel** [comment]
Do you understand what API means? It means "Application Programming Interface". It's an interface that you use via programming, that does stuff with an application.

Godot has an API, namely the [**class API**](http://docs.godotengine.org/en/stable/classes/_classes.html).
You use the *class API* to make Godot do stuff. (remember, it's "the engine uses your code, not the other way around")

Okay, so there's an API you can use to make Godot do stuff, for example build games. After all that's what Godot was made for.

class API = Godot API. Cool? Cool.

-----

Now to the two projects you're referring to. **I** am working on DLScript. I am *extending* Godot functionality with this module **as a part of Godot**. What DLScript does is: make *Godot's API* (the classes, remember?) available to shared libraries. It also loads these libraries and makes it so it's working well with the rest of Godot's scripting system.

DLScript didn't make "*their own API*", all it does it make *the existing* API available for use in shared libraries.

The mono module does the same, and it's not discontinued, it's being worked on, just not pushed to Github all the time. So, the mono module makes *the existing* API available for use in mono.

----

I'm pretty sure you're exaggerating about the C++ part. I can write C++ code reasonably fast and I don't spend 80% of my time debugging my or someone else's code. Considered switching away from Windows as your dev environment?

----

These two modules try to extend Godot in ways it was meant to be used. You want to use it in ways it wasn't meant to be used.

I already answered your original question: "Why Godot has no API?". The answer is, it has an [API](http://docs.godotengine.org/en/stable/classes/_classes).

What you're trying to do is use Godot in ways it never was meant to be used, and that's cool. Just don't expect upstream support for stuff that's against the "Godot way".

A knife manufacturer won't make their knifes softer because you like to use their knifes to slip into your shoes.

**TL;DR**: No

**Comment by CharlesWoodhill** [comment]
sorry i deleted that comment because i was angry when i wrote it, and i apologise for it.

**Comment by karroffel** [comment]
My point being: you have the source code, Godot is free, you can do with it whatever you want. Build tools the way you like it! Use the C++ classes (and with that breaking portability), you're free to do that.

You want to use Godot for something that Unity doesn't let you do? Perfect, it's free, do what you want. Just don't expect the devs to rebuild their whole system just so that *a veeeeeeeeery tiny* fraction of users would benefit from it.

Build modules in C++! You have all the power you want! It's nicely [documented](http://docs.godotengine.org/en/stable/reference/_developing.html) as well.

Godot does what it's meant to be, and it does that good IMO. For everything else you need to get your hands dirty.

**Comment by CharlesWoodhill** [comment]
i give up and close it, sorry again.
